From 70bd3323d18085299e54548f96d6a56b2223dd1a Mon Sep 17 00:00:00 2001
From: Matthijs van Duin <matthijsvanduin@gmail.com>
Date: Sat, 16 Aug 2025 10:01:19 +0200
Subject: [PATCH 16/16] spidev: undo misguided prohibition against
 compatible="spidev"

---
 drivers/spi/spidev.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/spi/spidev.c b/drivers/spi/spidev.c
index 653f82984216..5a4bea057189 100644
--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -729,6 +729,7 @@ static int spidev_of_check(struct device *dev)
 }
 
 static const struct of_device_id spidev_dt_ids[] = {
+	{ .compatible = "spidev" },
 	{ .compatible = "cisco,spi-petra", .data = &spidev_of_check },
 	{ .compatible = "dh,dhcom-board", .data = &spidev_of_check },
 	{ .compatible = "elgin,jg10309-01", .data = &spidev_of_check },
-- 
2.39.5

